lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1360 @@
+        if (!OpWidth)
+          OpWidth = Width;
+      Comps.push_back(Comp);
+    }
+    for (auto Const : InitConstComps) {
+      if (!Const.Width)
+        Const.Width = Width;
+      ConstComps.push_back(Const);
+    }
+  }
+  // Third, create input/constant component(s) for the output width
+  if (!InputWidths.count(LHS->Width)) {
+    if (InputWidths.size()) {
+      for (auto Comp : InitComps) {
+        if (Comp.Width && Comp.Width != LHS->Width)
+          continue;
+        if (!Comp.Width)
+          Comp.Width = LHS->Width;
+        for (auto &OpWidth : Comp.OpWidths)
+          if (!OpWidth)
+            OpWidth = LHS->Width;
+        Comps.push_back(Comp);
+      }
+    }
+    for (auto Const : InitConstComps) {
+      Const.Width = LHS->Width;
+      ConstComps.push_back(Const);
+    }
+  }
+}
+
+void InstSynthesis::initInputVars(InstContext &IC) {
+  std::vector<Inst *> Tmp;
+  getInputVars(LHS, Tmp);
+  // Remove duplicates
+  std::set<Inst *> TmpSet;
+  for (auto I : Tmp)
+    if (TmpSet.insert(I).second) {
+      Inputs.push_back(I);
+      InputWidths.insert(I->Width);
+    }
+  for (unsigned J = 0; J < Inputs.size(); ++J) {
+    // Note that location variable 0_0 is not used
+    LocVar In = std::make_pair(0, J+1);
+    std::string LocVarStr = getLocVarStr(In, LOC_PREFIX);
+    Inst *Loc = IC.createVar(LocInstWidth, LocVarStr);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Add input location to I
+    I.emplace_back(In, Loc);
+    // Update input name
+    LocVarStr = getLocVarStr(In, INPUT_PREFIX);
+    Inputs[J]->Name = LocVarStr;
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Update CompInstMap map with concrete Inst
+    CompInstMap[In] = Inputs[J];
+  }
+}
+
+void InstSynthesis::initConstComponents(InstContext &IC) {
+  for (unsigned J = 0; J < ConstComps.size(); ++J) {
+    auto const &Comp = ConstComps[J];
+    // Treat constant comps as ordinary inputs
+    LocVar In = std::make_pair(0, Inputs.size()+J+1);
+    std::string LocVarStr = getLocVarStr(In, LOC_PREFIX);
+    Inst *Loc = IC.createVar(LocInstWidth, LocVarStr);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Add constant component location to I
+    I.emplace_back(In, Loc);
+    // Update name
+    LocVarStr = getLocVarStr(In, CONST_PREFIX);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Update CompInstMap map with concrete Inst
+    assert(Comp.Width && "const comp width not set");
+    Loc = IC.createVar(Comp.Width, LocVarStr);
+    CompInstMap[In] = Loc;
+  }
+}
+
+void InstSynthesis::filterFixedWidthIntrinsicComps() {
+  // CtPop/BSwap/Cttz/Ctlz require specific widths
+  for (auto It = Comps.begin(); It != Comps.end();) {
+    assert(It->Width && "comp width not set");
+    if (It->Kind == Inst::BSwap && It->Width != 16 && It->Width != 32 &&
+        It->Width != 64)
+      It = Comps.erase(It);
+    else if (((It->Kind == Inst::CtPop) || (It->Kind == Inst::Ctlz) ||
```

Comment:
personally I'd put more curly braces in a collection of ifs/elses that is this complicated